### ISSUE TITLE
[BUG] Unable to specify existing app certificate from different resource group than web app

### ISSUE THREAD

**Issue by tostringtheory** [Azure/azure-libraries-for-net#961]
**Describe the bug**
When attempting to update a WebApp and define a SslBinding using an existing certificate, that does not exist in the same resource group as the webapp, the operation fails with a `System.NullReferenceException: 'Object reference not set to an instance of an object.'`.

***Exception or Stack Trace***
```
	This exception was originally thrown at this call stack:
	Microsoft.Azure.Management.AppService.Fluent.HostNameSslBindingImpl<FluentT, FluentImplT, DefAfterRegionT, DefAfterGroupT, UpdateT>.NewCertificateAsync.AnonymousMethod__40_0()
	System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
	System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
	System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
	System.Runtime.CompilerServices.TaskAwaiter<TResult>.GetResult()
	Microsoft.Azure.Management.AppService.Fluent.WebAppBaseImpl<FluentT, FluentImplT, DefAfterRegionT, DefAfterGroupT, UpdateT>.CreateResourceInternalAsync(System.Threading.CancellationToken)
	System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
	System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
	System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
	System.Runtime.CompilerServices.TaskAwaiter<TResult>.GetResult()
    ...
    [Call Stack Truncated]
```

**To Reproduce**
Create a resource group (1)
Create an app service plan in 1
Create an app service certificate in 1

Create another resource group (2)
Create a web app in 2
Attempt to add a hostname binding to the webapp utilizing the app service certificate from 1

***Code Snippet***
```
		private async Task ConfigureSsl(IWebApp appService, string fqdn)
		{
			using (logger.BeginScope("Configuring SSL for {fqdn}", fqdn))
			{
				logger.LogTrace("Finding certificate for appservice");
				var certificates = (await hosts.Certificates);
				var certificate = certificates.Single(c => c.HostNames.Contains(configuration.Company.Sso.Zone));



				logger.LogDebug("Creating SSL binding for {fqdn} to {certificate-name}", fqdn, certificate.Name);
				await appService.Update()
								.DefineSslBinding()
								.ForHostname(fqdn)
								.WithExistingCertificate(certificate.Thumbprint)
								.WithSniBasedSsl()
								.Attach()
								.ApplyAsync();
			}
		}
```

**Expected behavior**
I would expect that the SDK accepts the certificate name or thumbprint as specified.  The issue here is that the code for the method `WithExistingCertificate` lists's all certificates to probe/find one that matches the input parameter.  This match utilizes the Inner manager for the webapp, which ends up scoping the resources to the resource group that the app belongs to.  However, the app service via the portal and REST api is perfectly happy to accept a certificate in a different group.

**Screenshots**
Not-available

**Setup (please complete the following information):**
 - OS: Windows 10 19041.21
 - IDE : Visual Studio 2019 16.4.2
 - Version of the Library used : 1.30.0

**Additional context**
I was able to work around this method by adding an extension method to encapsulate some reflection and mimic the behavior of the method, while supplanting more direct logic:

```
	public static class CertificateDefinitionExtensions
	{
		//shimming this method: https://github.com/Azure/azure-libraries-for-net/blob/5b011ed7fd630b55a20906ffd97d65272312a18d/src/ResourceManagement/AppService/HostNameSslBindingImpl.cs#L160
		//this is due to the fact that the implementation defined there will only look in the local resource group for  the certificate
		public static IWithSslType<T> WithExistingCertificate<T>(this IWithCertificate<T> fluent, IAppServiceCertificate certificate)
		{
			var type = fluent.GetType();
			var newCertificateField = type.GetField("newCertificate", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.SetField);
			var withCertificateThumbprintMethod = type.GetMethod("WithCertificateThumbprint", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.InvokeMethod);

			Func<Task<IAppServiceCertificate>> newCertificate = () =>
			{
				withCertificateThumbprintMethod.Invoke(fluent, new object[] { certificate.Thumbprint });

				return Task.FromResult(certificate);
			};

			newCertificateField.SetValue(fluent, newCertificate);

			return (IWithSslType<T>)fluent;
		}
	}
```

This works perfectly, however I hate reflecting into 3rd party assemblies for obvious reasons.  Ideally, the fluent interface would just allow you to pass in the requested IAppServiceCertificate like I had and perform the same operation.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [√] Bug Description Added
- [√] Repro Steps Added
- [√] Setup information Added

**Comment by weidongxu-microsoft** [comment]
This appears to be same bug as https://github.com/Azure/azure-libraries-for-net/issues/917

That one was fixed, but not caught in recent 1.30 release.

**Comment by ChenTanyi** [comment]
@weidongxu-microsoft @tostringtheory this will be fixed in 1.31.0 rather than 1.30.0.

**Comment by yaohaizh** [comment]
Will fixed on 1.31.0.

**Comment by steeve49** [comment]
@tostringtheory Ok I think I found why it doesn't work...
I thought that using the WithExistingCertificate command would do the "installation" of existing certificate into my newly created resource group but it doesn't.
I just looked at my ARM template workaround to see what I have to find into azure sdk fluent to do the same thing but I don't know how to find it.

`{
      "type": "Microsoft.Web/certificates",
      "name": "MyCertificate",
      "apiVersion": "2018-11-01",
      "location": "[variables('Location')]",
      "properties": {
        "keyVaultId": "[parameters('keyvaultId')]",
        "keyVaultSecretName": "[parameters('KeyvaultSecretName')]",
        "thumbprint": "[parameters('certThumbprint')]"
      }
    }`

Do you know which function I have to use to do the same thing with SDK Fluent ?

**Comment by steeve49** [comment]
@weidongxu-microsoft Thanks for having merged the two issues.
As I told into #6409 I really don't know if my process miss a step or not, you can find details in my last answer in #6409 topic.
I have tried many other options but nothing works (but my ARM template workaround at the moment).

I also forgot to answer to @tostringtheory about fiddler. I've tried and what I receive is a 404 error not found.

Request Json : 

`{
  "properties": {
    "enabled": true,
    "hostNameSslStates": [
      {
        "name": "myname-tests-web-gef.azurewebsites.net",
        "sslState": "Disabled",
        "hostType": "Standard"
      },
      {
        "name": "mysubdomain.mydomain.fr",
        "sslState": "SniEnabled",
        "thumbprint": "myAlreadyInstalledCertificateThumbprint",
        "toUpdate": true
      },
      {
        "name": "myname-tests-web-gef.scm.azurewebsites.net",
        "sslState": "Disabled",
        "hostType": "Repository"
      }
    ],
    "serverFarmId": "/subscriptions/mySubscriptionId/resourceGroups/myname-tests-rg/providers/Microsoft.Web/serverfarms/myname-tests-appplan",
    "reserved": false,
    "isXenon": false,
    "hyperV": false,
    "siteConfig": {},
    "scmSiteAlsoStopped": false,
    "clientAffinityEnabled": true,
    "clientCertEnabled": false,
    "hostNamesDisabled": false,
    "containerSize": 0,
    "dailyMemoryTimeQuota": 0,
    "httpsOnly": false,
    "redundancyMode": "None"
  },
  "kind": "app",
  "location": "West Europe",
  "tags": {}
}`

The response Json :

`{"Code":"NotFound","Message":"Certificate MyThumbPrint was not found.","Target":null,"Details":[{"Message":"Certificate MyThumbPrint was not found."},{"Code":"NotFound"},{"ErrorEntity":{"ExtendedCode":"04031","MessageTemplate":"Certificate {0} was not found.","Parameters":["MyThumbPrint "],"Code":"NotFound","Message":"Certificate MyThumbPrint was not found."}}],"Innererror":null}`

**Comment by weidongxu-microsoft** [comment]
@steeve49 

The ARM template seems for creating a certificate, not for binding it to a web app (no property relate to web app, and why keyvaultId and keyvaultSecretName, anything stored in keyvault?).

It might help if you can share the task you want to achieve, and what exists on your subscription (a screenshot of Portal?), and what is your current code. Is it any scenario mentioned in https://docs.microsoft.com/en-us/azure/app-service/configure-ssl-certificate?

**Comment by steeve49** [comment]
Ok, so here is my backup resource group where I have bought an azure ssl wildcard certificate and stored its private key into the keyvault as your link says here : https://docs.microsoft.com/en-us/azure/app-service/configure-ssl-certificate#store-in-azure-key-vault

![image](https://user-images.githubusercontent.com/61324521/84366034-95078000-abd2-11ea-8094-d1cc29da8a03.png)

I have some deployment console applications that use Azure SDK Fluent to create resource groups and resources (webapps, functions, etc). In these resources I have the need to bind my SSL certificate to my webapps for example.

As you can see here, my certificate is already "bound" to three different resourcegroups using my method (I will explain just after)

![image](https://user-images.githubusercontent.com/61324521/84366517-3b538580-abd3-11ea-8fc2-2026fd642a4a.png)

Now, here is how I make it work using azure pipelines :
I have two deployment tasks :
- Azure resource group deployment (this is the workaround key)
- Azure CLI

The azure resource group deployment task creates my new resource group using a really light ARM template that create an app plan and make the stuff with certificate (because I don't really know what it does) using Keyvault informations and existing certificate thumbprint.

`{
      "type": "Microsoft.Web/certificates",
      "name": "AlreadyExistingCertificateName",
      "apiVersion": "2018-11-01",
      "location": "[variables('Location')]",
      "properties": {
        "keyVaultId": "[parameters('keyvaultId')]",
        "keyVaultSecretName": "[parameters('KeyvaultSecretName')]",
        "thumbprint": "[parameters('certThumbprint')]"
      },
 }`

Now, the CLI Task execute my deployment process using the fluent SDK.

I tried to do the minimum part : 

1. Get or create the resource group
2. Get or create the appPlan
3. Create a new WebApp
4. Create my hostnamebinding and bind my ssl certificate

And now let's focus to the ssl part : 
After creating the webApp, I run this code

`await webApp.Update()
                                 .DefineHostnameBinding()
                                     .WithThirdPartyDomain(domain)
                                     .WithSubDomain(subdomain)
                                     .WithDnsRecordType(CustomHostNameDnsRecordType.CName)
                                 .Attach()
                                 .DefineSslBinding()
                                    .ForHostname($"{subdomain}.{domain}")
                                    .WithExistingCertificate(certificateThumbprint)
                                    .WithSniBasedSsl()
                                 .Attach()
                                 .ApplyAsync();` 

When the Azure resource group deployment task is executed using ARM tempate => it works.
When I remove this task, it throws a notfound exception.

Here is a screenshot of my app after the crash

![image](https://user-images.githubusercontent.com/61324521/84367826-0fd19a80-abd5-11ea-94de-09a9103266a7.png)

As you can see, my custom domain is here but not secured with SSL certificate.
If I click on "add binding" it asks me to select and import or upload a certificate.

![image](https://user-images.githubusercontent.com/61324521/84367970-46a7b080-abd5-11ea-8354-0102a8edd39b.png)

I click on "Import App Service Certificate"
In the next screenshot, you can see that my already existing certificate (stored into my backup resource group using keyvault etc...) is here and valid

![image](https://user-images.githubusercontent.com/61324521/84368109-6f2faa80-abd5-11ea-86a9-917bb169f4db.png)

![image](https://user-images.githubusercontent.com/61324521/84368331-b61da000-abd5-11ea-8d76-8f85a285cd54.png)

I validate the certificate and then my app is bound to the SSL certificate and secured

![image](https://user-images.githubusercontent.com/61324521/84368392-cf265100-abd5-11ea-9bf0-5a015c13e9ab.png)

**My need :** 
I want to remove the Azure resource group deployment task which depends on ARM template to fully run my deployment through Azure Sdk Fluent application.

**Comment by tostringtheory** [comment]
@steeve49 - I got ya!  When I was working on our deployments I hit against the same limitation, the FluentSDK (at least at the time I was working), didn't allow you to fluently specify/create `IAppServiceCertificate`'s from KeyVault.  Under the covers though, everything in fluent is essentially calling the ARM endpoints anyways, so I poked around in the SDK until I found what I needed.  There's a type that has all of the necessary plumbing - the `IResourceManager`.  I'll preface this by saying I don't know if this has been made easier to access, but at the time that I was writing my deployment code, it wasn't, so reflection inbound.

FIRST - I amended the `IAzure` interface to include the useful type (I'd love it if this was added to the base `IAzure` type):

```
	public interface IAzureSubscription
		: IAzure
	{
		IResourceManager ResourceManager { get; }
	}
```
Next, we need to implement this class, and create an `IResourceManager` instance.  To create an `IResourceManager` instance, the framework wants a `RestClient`.  Great.  The `RestClient` that the Fluent SDK uses is in a private field of the `IAuthenticated` type, so I had to pull the values out that way.  The rest of the class is essentially just proxying the calls to the original `IAzure` instance (you need a lot of usings):

```
using Microsoft.Azure.Management.AppService.Fluent;
using Microsoft.Azure.Management.BatchAI.Fluent;
using Microsoft.Azure.Management.Cdn.Fluent;
using Microsoft.Azure.Management.Compute.Fluent;
using Microsoft.Azure.Management.ContainerInstance.Fluent;
using Microsoft.Azure.Management.ContainerRegistry.Fluent;
using Microsoft.Azure.Management.ContainerService.Fluent;
using Microsoft.Azure.Management.CosmosDB.Fluent;
using Microsoft.Azure.Management.Dns.Fluent;
using Microsoft.Azure.Management.Eventhub.Fluent;
using Microsoft.Azure.Management.Fluent;
using Microsoft.Azure.Management.KeyVault.Fluent;
using Microsoft.Azure.Management.Locks.Fluent;
using Microsoft.Azure.Management.Monitor.Fluent;
using Microsoft.Azure.Management.Msi.Fluent;
using Microsoft.Azure.Management.Network.Fluent;
using Microsoft.Azure.Management.PrivateDns.Fluent;
using Microsoft.Azure.Management.Redis.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent.Core;
using Microsoft.Azure.Management.Search.Fluent;
using Microsoft.Azure.Management.ServiceBus.Fluent;
using Microsoft.Azure.Management.Sql.Fluent;
using Microsoft.Azure.Management.Storage.Fluent;
using Microsoft.Azure.Management.TrafficManager.Fluent;
using Microsoft.Rest.Azure;
using System.Collections.Generic;
using System.Reflection;
```
And the implementation:
```
public class AzureSubscription
	: IAzureSubscription
{
	public AzureSubscription(IAzure azure)
	{
		this.azure = azure;
		this.authenticated = (Azure.IAuthenticated)azure.GetType().GetField("authenticated", BindingFlags.GetField | BindingFlags.NonPublic | BindingFlags.Instance).GetValue(azure);
		this.restClient = (RestClient)authenticated.GetType().GetField("restClient", BindingFlags.GetField | BindingFlags.NonPublic | BindingFlags.Instance).GetValue(authenticated);

		this.ResourceManager = Microsoft.Azure.Management.ResourceManager.Fluent.ResourceManager.Authenticate(this.restClient)
			.WithSubscription(azure.SubscriptionId);
	}

	private readonly IAzure azure;
	private readonly Azure.IAuthenticated authenticated;
	private readonly RestClient restClient;

	public IResourceManager ResourceManager { get; }

	public Microsoft.Azure.Management.ResourceManager.Fluent.ISubscription GetCurrentSubscription() => this.azure.GetCurrentSubscription();
	public IEnumerable<IAzureClient> ManagementClients => this.azure.ManagementClients;
	public string SubscriptionId => this.azure.SubscriptionId;
	public IServiceBusNamespaces ServiceBusNamespaces => this.azure.ServiceBusNamespaces;
	public ILoadBalancers LoadBalancers => this.azure.LoadBalancers;
	public IAccessManagement AccessManagement => this.azure.AccessManagement;
	public IApplicationGateways ApplicationGateways => this.azure.ApplicationGateways;
	public Microsoft.Azure.Management.ResourceManager.Fluent.ISubscriptions Subscriptions => this.azure.Subscriptions;
	public IResourceGroups ResourceGroups => this.azure.ResourceGroups;
	public IStorageAccounts StorageAccounts => this.azure.StorageAccounts;
	public IVirtualMachines VirtualMachines => this.azure.VirtualMachines;
	public IVirtualMachineScaleSets VirtualMachineScaleSets => this.azure.VirtualMachineScaleSets;
	public INetworks Networks => this.azure.Networks;
	public INetworkSecurityGroups NetworkSecurityGroups => this.azure.NetworkSecurityGroups;
	public IPublicIPAddresses PublicIPAddresses => this.azure.PublicIPAddresses;
	public INetworkInterfaces NetworkInterfaces => this.azure.NetworkInterfaces;
	public IPrivateLinkServices PrivateLinkServices => this.azure.PrivateLinkServices;
	public IAzureFirewalls AzureFirewalls => this.azure.AzureFirewalls;
	public IRouteTables RouteTables => this.azure.RouteTables;
	public IDeployments Deployments => this.azure.Deployments;
	public IPolicyAssignments PolicyAssignments => this.azure.PolicyAssignments;
	public IPolicyDefinitions PolicyDefinitions => this.azure.PolicyDefinitions;
	public IVirtualMachineImages VirtualMachineImages => this.azure.VirtualMachineImages;
	public IVirtualMachineExtensionImages VirtualMachineExtensionImages => this.azure.VirtualMachineExtensionImages;
	public IAvailabilitySets AvailabilitySets => this.azure.AvailabilitySets;
	public IVaults Vaults => this.azure.Vaults;
	public ITrafficManagerProfiles TrafficManagerProfiles => this.azure.TrafficManagerProfiles;
	public IDnsZones DnsZones => this.azure.DnsZones;
	public IPrivateDnsZones PrivateDnsZones => this.azure.PrivateDnsZones;
	public ISqlServers SqlServers => this.azure.SqlServers;
	public IRedisCaches RedisCaches => this.azure.RedisCaches;
	public ICdnProfiles CdnProfiles => this.azure.CdnProfiles;
	public IVirtualMachineCustomImages VirtualMachineCustomImages => this.azure.VirtualMachineCustomImages;
	public IDisks Disks => this.azure.Disks;
	public ISnapshots Snapshots => this.azure.Snapshots;
	public IComputeSkus ComputeSkus => this.azure.ComputeSkus;
	public IEventHubNamespaces EventHubNamespaces => this.azure.EventHubNamespaces;
	public IEventHubs EventHubs => this.azure.EventHubs;
	public IEventHubDisasterRecoveryPairings EventHubDisasterRecoveryPairings => this.azure.EventHubDisasterRecoveryPairings;
	public IAlertRules AlertRules => this.azure.AlertRules;
	public IAutoscaleSettings AutoscaleSettings => this.azure.AutoscaleSettings;
	public IRegistryTasks ContainerRegistryTasks => this.azure.ContainerRegistryTasks;
	public IRegistryTaskRuns RegistryTaskRuns => this.azure.RegistryTaskRuns;
	public INetworkWatchers NetworkWatchers => this.azure.NetworkWatchers;
	public IVirtualNetworkGateways VirtualNetworkGateways => this.azure.VirtualNetworkGateways;
	public ILocalNetworkGateways LocalNetworkGateways => this.azure.LocalNetworkGateways;
	public IExpressRouteCircuits ExpressRouteCircuits => this.azure.ExpressRouteCircuits;
	public IApplicationSecurityGroups ApplicationSecurityGroups => this.azure.ApplicationSecurityGroups;
	public IRouteFilters RouteFilters => this.azure.RouteFilters;
	public IDdosProtectionPlans DdosProtectionPlans => this.azure.DdosProtectionPlans;
	public IWebApps WebApps => this.azure.WebApps;
	public IAppServiceManager AppServices => this.azure.AppServices;
	public ISearchServices SearchServices => this.azure.SearchServices;
	public IContainerServices ContainerServices => this.azure.ContainerServices;
	public IKubernetesClusters KubernetesClusters => this.azure.KubernetesClusters;
	public ICosmosDBAccounts CosmosDBAccounts => this.azure.CosmosDBAccounts;
	public IContainerGroups ContainerGroups => this.azure.ContainerGroups;
	public IRegistries ContainerRegistries => this.azure.ContainerRegistries;
	public IManagementLocks ManagementLocks => this.azure.ManagementLocks;
	public IIdentities Identities => this.azure.Identities;
	public IBatchAIWorkspaces BatchAIWorkspaces => this.azure.BatchAIWorkspaces;
	public IBatchAIUsages BatchAIUsages => this.azure.BatchAIUsages;
	public IActivityLogs ActivityLogs => this.azure.ActivityLogs;
	public IMetricDefinitions MetricDefinitions => this.azure.MetricDefinitions;
	public IDiagnosticSettings DiagnosticSettings => this.azure.DiagnosticSettings;
	public IActionGroups ActionGroups => this.azure.ActionGroups;
	public IGalleries Galleries => this.azure.Galleries;
	public IGalleryImages GalleryImages => this.azure.GalleryImages;
	public IGalleryImageVersions GalleryImageVersions => this.azure.GalleryImageVersions;
}
```
Access to the `IResourceManager` now allows you to declare, Fluently, any ARM type.  In this case, here's my code that creates an app service certificate from a keyvault reference.  You can see how closely it resembles your ARM template:
```
logger.LogDebug("Creating app service certificate");
var resource = await resources.ResourceManager.GenericResources
	.Define(certificateName)
	.WithRegion(Region.USSouthCentral)
	.WithExistingResourceGroup(resourceGroup)
	.WithResourceType("certificates")
	.WithProviderNamespace("Microsoft.Web")
	.WithoutPlan()
	.WithApiVersion("2019-08-01")
	.WithProperties(JObject.Parse(JsonConvert.SerializeObject(new
	{
		KeyVaultId = vault.Id,
		KeyVaultSecretName = definition.Certificate,
		ServerFarmId = appServicePlan.Id
	})))
	.CreateAsync();
```
Besides the rest of the ick, I'll point out the specific issue I faced with the `WithProperties` call.  When I tried to just send the object, without the serialize/parse, the sdk rejected the request (can't remember the exact reason at the moment.  But If I gave it a `JObject`,  then the SDK just ferried that along happily.

I like parts of the fluent SDK, but found that while they had several components implemented, they were lacking others that we needed, and the support/responsiveness wouldn't meet our demands long term.  I've actually begun transitioning to just directly using the ARM REST API without ARM templates or fluent SDK.  A minimal wrapper, that does the calls, and that's it.  It's got a lot less friction, and more of a guarantee to work since it's the same API that everything else (CLI, Portal, SDK, etc) uses.  In cases where I'm facing an issue I can look at the request formulated in the portal/CLI, and figure out the difference between my REST call.  

I tried to make sure I had everything in this post to implement this in your code, if you find something's missing, let me know! Hope that helps.

**Comment by Patafoin** [comment]
@tostringtheory : You are my hero ! It works perfectly !!!!! Many many many thanks to you ! Really, It's amazing ! It's been about 2 years I search about a real good solution and no one in microsoft or stackoverflow could help me. Thanks a lot !

I think it's really strange that no one felt into the same trap and posted something on the web about this solution. If you have a personal webpage, you should post it ^^ it's a great help !

@weidongxu-microsoft Do you know if actual library version can do the same thing without using a custom class ? 

I can sleep happy now :)

**Comment by tostringtheory** [comment]
@Patafoin - glad to hear that's working for you!  One of the other issues (#27 ) I also worked around in the same way.  Having access to create any un-supported resources by the fluent interface (at least partially) is useful:

```
var resource = await resources.ResourceManager.GenericResources
	.Define(nameBuilder.ForRoleInstance("insights"))
	.WithRegion(Region.USSouthCentral)
	.WithExistingResourceGroup(resourceGroup)
	.WithResourceType("components")
	.WithProviderNamespace("Microsoft.Insights")
	.WithoutPlan()
	.WithApiVersion("2015-05-01")
	.WithProperties(new
	{
		Flow_Type = "Bluefield",
		Application_Type = "web",
		Request_Source = "rest"
	}).CreateAsync();
```
What perplexes me looking at this code, is that I DON'T see the same serialize/deserialize step that I had to do in the certificates function, so I don't know/remember why I had to do that.  I remember I had some exception at the time that led me into making it a `JObject`, but don't remember what.  Maybe it's resolved?

As for no one falling into the trap, I think a part of that might be peoples aversion (including mine) to reflection of libraries internal functions (like above)...  In my case, I just have some unit tests elsewhere confirming the reflections working separately of the execution.

**Comment by weidongxu-microsoft** [comment]
`GenericResource` can be get from `IAzure` by `azure.Deployments.Manager.GenericResources` (me too in surprise that this is not exposed more directly). `azure.Deployments.Manager` is the `IResourceManager`.

Yes, this is a more generic approach for scenario currently not well supported by Fluent SDK.

**Comment by weidongxu-microsoft** [comment]
@steeve49 
Thanks for the description. It is clear.

I currently do not have a wildcard SSL to test.
Following code might work, if you can find your SSL in `certificateOrders` returned in 1st line:

```
            IEnumerable<IAppServiceCertificateOrder> certificateOrders = azure.AppServices.AppServiceCertificateOrders.List();
            IAppServiceCertificateOrder certificateOrder = certificateOrders.First();

            azure.AppServices.AppServiceCertificates.Define("certificateName")
                .WithRegion(Region.USEast)
                .WithExistingResourceGroup(RESOURCE_GROUP)
                .WithExistingCertificateOrder(certificateOrder)
                .Create();
```

**Comment by Patafoin** [comment]
@weidongxu-microsoft : your method throws a "Microsoft.Rest.ValidationException: ''Password' cannot be null.'". I already tried this one but it doesn't work unfortunately.

@tostringtheory Thanks for the application insights snippet, it's really interesting because I faced the same problem that I solved using rest calls. (why I didn't do that for certificate ? because I didn't knew where to search a solution :) and didn't knew there was an extra step to make my certificate available into my resource group)
